lib/apipie_bindings/action.rb

```diff
@@ -57,35 +55,40 @@ def find_route(params={})
       return suitable_route
     end
 
-    def validate!(params)
-      missing_params = missing_params(api_params_tree { |par| par.required? }, params_tree(params))
-      raise ApipieBindings::MissingArgumentsError.new(missing_params) unless missing_params.empty?
-    end
-
-    def api_params_tree(&block)
-      ApipieBindings::Utilities.params_hash_tree(self.params, &block)
-    end
-
-    def params_tree(params)
-      params.inject([]) do |tree, (key, val)|
-        subtree = val.is_a?(Hash) ? { key.to_s => params_tree(val) } : key.to_s
-        tree << subtree
-        tree
-      end
+    def validate!(parameters)
+      errors = validate(params, parameters)
+      raise ApipieBindings::MissingArgumentsError.new(errors) unless errors.empty?
     end
 
-    def missing_params(master, slave)
-      missing = []
-      master.each do |required_param|
-        if required_param.is_a?(Hash)
-          key = required_param.keys.first
-          slave_hash = slave.select { |p| p.is_a?(Hash) && p[key] }
-          missing << missing_params(required_param[key], slave_hash.first ? slave_hash.first[key] : [])
-        else
-          missing << required_param unless slave.include?(required_param)
+    def validate(level_params, level_given, path='%s')
```

Comment:
what about

```ruby
def validate(params, values, path)
```

When reading for the first time and seeing the level there, I was assuming the level_given would contain some number indicating the level of depth.